diff -ur a/company-tabnine.el b/company-tabnine.el
--- a/company-tabnine.el	1970-01-01 03:00:01.000000000 +0300
+++ b/company-tabnine.el	2021-04-22 14:34:13.329024811 +0300
@@ -376,7 +376,7 @@
            :sentinel #'company-tabnine--process-sentinel
            :noquery t)))
   ;; hook setup
-  (message "TabNine server started.")
+  ;; (message "TabNine server started.")
   (dolist (hook company-tabnine--hooks-alist)
     (add-hook (car hook) (cdr hook))))

@@ -475,17 +475,17 @@
   (when (and company-tabnine--process
              (memq (process-status process) '(exit signal)))

-    (message "TabNine process %s received event %s."
-             (prin1-to-string process)
-             (prin1-to-string event))
+    ;; (message "TabNine process %s received event %s."
+             ;; (prin1-to-string process)
+             ;; (prin1-to-string event))

     (if (>= company-tabnine--restart-count
             company-tabnine-max-restart-count)
         (progn
-          (message "TabNine process restart limit reached.")
+          ;; (message "TabNine process restart limit reached.")
           (setq company-tabnine--process nil))

-      (message "Restarting TabNine process.")
+      ;; (message "Restarting TabNine process.")
       (company-tabnine-start-process)
       (setq company-tabnine--restart-count
             (1+ company-tabnine--restart-count)))))

